src/terrain/clients/iplant_groups.clj

```diff
@@ -57,3 +60,49 @@
     (when-not (#{200 404} status)
       (throw (Exception. (str "iplant-groups service returned status " status))))
     {:subjects (:subjects (:body res))}))
+
+(defn- get-client []
+  (c/new-cyverse-groups-client (config/ipg-base) (config/environment-name)))
+
+(defn- get-optional-folder [client user name]
+  (try+
+   (c/get-folder client user name)
+   (catch [:status 404] _ nil)))
+
+(defn- create-user-folder [client user name]
+  (let [folder (c/add-folder client (config/grouper-user) name "")]
+    (c/grant-folder-privilege client (config/grouper-user) name user :stem)
+    folder))
+
+(defn- get-or-add-folder [client user name]
+  (or (get-optional-folder client user name)
+      (create-user-folder client user name)))
+
+(defn- get-collaborator-list-folder [user client]
+  (let [name (c/build-folder-name client (format "users:%s:collaborator-lists" user))]
+    (get-or-add-folder client user (string/replace name  #":[^:]+$" ""))
+    (get-or-add-folder client user name)))
+
```

Comment:
Your understanding of this is correct.

I tend to use `or` for things like this quite a bit especially when external API calls are involved, but your changes make sense to me as well. I'll make these changes.